cirq_google.GridDevice, minus gateset and gate durations

verult opened the pull request:
Part of https://github.com/quantumlib/Cirq/issues/5050

* Implemented qubit set, qubit pairs, and op validation. Gateset and gate duration will be done in a follow-up.
* Added a `device_specification_validation` module shell, to be used by the QCS server side to validate the DeviceSpecification proto against things like: 
  * Qubit self-loops in qubit pairs
  * Qubit pairs contain qubits not in the valid qubit set.
  * Gate durations contains a gate which is missing in the gateset.

`__str__()` and `_repr_pretty_` are nearly all copied from `SerializableDevice`.

Part of #5050

Also fixes #5197

**Question:** Does this definition of device equality make sense, or should it contain device name as well? Implementing equality is convenient for json and repr tests but not sure if it make sense in general.

@dstrain115 @MichaelBroughton

verult commented:
Should exceptions need to have repr and json files?

MichaelBroughton commented on cirq-core/cirq/devices/grid_device_metadata.py:
```diff
@@ -169,4 +169,9 @@ def _json_dict_(self):
     @classmethod
     def _from_json_dict_(cls, qubit_pairs, gateset, gate_durations, all_qubits, **kwargs):
-        return cls(qubit_pairs, gateset, dict(gate_durations), all_qubits)
+        return cls(
+            qubit_pairs,
+            gateset,
+            None if gate_durations is None else dict(gate_durations),
+            all_qubits,
+        )
```
Why add this `None if gate durations` ?

verult replied:
Non-default arguments can't be after default arguments. Will shifting the order of method parameters here break JSON backward compatibility?

MichaelBroughton commented on cirq-core/cirq/ops/gateset.py:
```diff
@@ -406,9 +406,10 @@ def _value_equality_values_(self) -> Any:
     def __repr__(self) -> str:
         name_str = f'name = "{self.name}", ' if self.name is not None else ''
+        gates_str = f'{self._gates_repr_str}, ' if len(self._gates_repr_str) > 0 else ''
         return (
             f'cirq.Gateset('
-            f'{self._gates_repr_str}, '
+            f'{gates_str}'
```
Why change here ?

tanujkhattar replied:
Please add a test to `gateset_test.py`. Also, If the change is unrelated and is not blocking this PR, then I'd prefer if we can split it out in a separate PR.

verult replied:
Added and moved to https://github.com/quantumlib/Cirq/pull/5322

MichaelBroughton commented on cirq-google/cirq_google/devices/google_device.py:
```diff
@@ -0,0 +1,226 @@
+
+
+@cirq.value_equality
+class GoogleDevice(cirq.Device):
```
Might want to call it something like `UniformGoogleDevice` or `GoogleGridDevice` just to convey the message that this type identifies and works for devices that have uniform gates in all directions etc. Since we are surfacing a GridDeviceMetadata here we want to ensure that the user knows they won't get a device that doesn't have the same gates and connectivity everywhere.

verult replied:
Sounds good. How about just `GridDevice`? I'm cautious with "uniform" because for example some 2-qubit gates may be missing. The "Google" descriptor is implicit from the package name and was there only because I couldn't think of a better descriptor :)

MichaelBroughton commented on cirq-google/cirq_google/devices/google_device.py:
```diff
@@ -0,0 +1,226 @@
+
+    For Google devices, the
+    [DeviceSpecification proto](
+        https://github.com/quantumlib/Cirq/blob/3969c2d3964cea56df33b329f036ba6810683882/cirq-google/cirq_google/api/v2/device.proto#L13
+    )
+    is the main specification for device information surfaced by the Quantum Computing Service.
+    Thus, this class is should be instantiated using a `DeviceSpecification` proto via the
+    `from_proto()` class method.
```
Love the verbosity here!

MichaelBroughton commented on cirq-google/cirq_google/devices/google_device.py:
```diff
@@ -0,0 +1,226 @@
+        """
+        self._metadata = metadata
+
+    @classmethod
+    def from_proto(cls, proto: v2.device_pb2.DeviceSpecification) -> 'GoogleDevice':
```
Why is the `from_proto` method preffered over init design-wise ?

verult replied:
The full DeviceSpecification would have to be stored in the repr and json, which could potentially be verbose?

MichaelBroughton commented on cirq-google/cirq_google/devices/google_device.py:
```diff
@@ -0,0 +1,226 @@
+        ]
+
+        # TODO(#5050) implement gate durations
+        try:
+            metadata = cirq.GridDeviceMetadata(
+                qubit_pairs=qubit_pairs,
+                gateset=cirq.Gateset(),  # TODO(#5050) implement
+                all_qubits=all_qubits,
+            )
+        except ValueError as ve:
+            raise ValueError("DeviceSpecification is invalid.") from ve
```
Nit: Do we want to catch the exception here and introduce this dependency on the behavior or GridDeviceMetaData's error behavior ? By extending our API for `from_proto` here to accomodate errors in downstream code we are kind of bringing the error checking from downstream code upstream here to some extent. Do we want to do that in this case ?

MichaelBroughton reviewed (commented):
Looking really good. Just a few minor tweaks then we should be good to merge.

verult reviewed (commented):
Thanks for the review!

verult commented:
Addressed comments, and added a validation function which is called at the beginning of `from_proto`, but also will eventually be called on server side prior to sending the `DeviceSpecification`.

dstrain115 commented on cirq-google/cirq_google/devices/grid_device.py:
```diff
@@ -0,0 +1,274 @@
+                        f" {target.ids}."
+                    )
+
+        # A SUBSET_PERMUTATION target should contain exactly one qubit.
```
I don't think this is true.  Why does a subset permutation contain exactly one qubit?

If I recall, a subset permutation means that you can have any subset of the qubits.  This is used for gates like measurement where you can measure any subset of the qubits.

verult replied:
SUBSET_PERMUTATION describes a *target set* rather than a target, where one can have any subset of the targets, with each target being exactly 1 qubit.

https://github.com/quantumlib/Cirq/blob/4ff626477bd93a81285c5673cc56044822eaa8af/cirq-google/cirq_google/api/v2/device.proto#L154-L159

dstrain115 replied:
Ah, I see.  Maybe link to this proto in the comment or otherwise clarify, since it is quite confusing to remember how this is done otherwise.

dstrain115 commented on cirq-google/cirq_google/devices/grid_device.py:
```diff
@@ -0,0 +1,274 @@
+    Example use cases:
+
+        * Get an instance of a Google grid device.
+        >>> device = cirq_google.get_engine().get_processor('weber').get_device()
```
I wonder if we should create an example that does not require access to the API.  This could be a follow-up issue/PR if this is not yet done.  For instance, using the virtual engine stuff that we are building.

dstrain115 commented on cirq-google/cirq_google/devices/grid_device.py:
```diff
@@ -0,0 +1,274 @@
+        >>> print(device)
+
+        * Determine whether a circuit can be run on the device.
+        >>> device.validate_circuit(circuit)  # Raises an exception if the circuit is invalid.
```
nit: may be more descriptive to say which exception is raised, e.g. ValueError so that people can put try catch around it easily.

dstrain115 commented on cirq-google/cirq_google/devices/grid_device.py:
```diff
@@ -0,0 +1,274 @@
+
+    For Google devices, the
+    [DeviceSpecification proto](
+        https://github.com/quantumlib/Cirq/blob/3969c2d3964cea56df33b329f036ba6810683882/cirq-google/cirq_google/api/v2/device.proto#L13
```
Change this like to point to master and not a specific commit.

dstrain115 commented on cirq-google/cirq_google/devices/grid_device.py:
```diff
@@ -0,0 +1,274 @@
+
+        This class only supports `cirq.GridQubit`s and `cirq.NamedQubit`s. If a
+        `DeviceSpecification.valid_qubits` string is in the form `<int>_<int>`, it is parsed as a
+        GridQubit. Otherwise it is parsed as a NamedQubit.
```
Interesting choice.  If this is a `GridDevice`, do you think the qubits should be restricted to GridQubits?  If there a use case for named qubits.  I would guess that it is likely a mistake if a non-grid qubit id happens.

verult replied:
Will also add validation for the GridQubit string format.

dstrain115 commented on cirq-google/cirq_google/devices/grid_device.py:
```diff
@@ -0,0 +1,274 @@
+            raise ValueError(f'Qubit pair is not valid on device: {operation.qubits!r}')
+
+    def __str__(self) -> str:
+        # If all qubits are grid qubits, render an appropriate text diagram.
```
I am surprised that the below code for rendering a grid device doesn't exist somewhere already.  Is the default insufficient because we do not assume pairwise connectivity between qubits?

verult replied:
I got this code from [`SerializableDevice`](https://github.com/quantumlib/Cirq/blob/4ff626477bd93a81285c5673cc56044822eaa8af/cirq-google/cirq_google/devices/serializable_device.py#L214), and there's also something similar for [`XmonDevice`](https://github.com/quantumlib/Cirq/blob/4ff626477bd93a81285c5673cc56044822eaa8af/cirq-google/cirq_google/devices/xmon_device.py#L223). This `GridDevice` will eventually replace both of them so I don't think there's a need to define a common function.

`cirq.Device` doesn't have a default, and I presume it's because yeah connectivity can be more general and would be hard to render. cc @MichaelBroughton

dstrain115 commented on cirq-google/cirq_google/devices/grid_device_test.py:
```diff
@@ -0,0 +1,228 @@
+    return grid_qubits, spec
+
+
+def _create_device_spec_qubit_pair_self_loops() -> v2.device_pb2.DeviceSpecification:
```
Optional:  consider adding a docstring for this and the below private functions for improved clarity.

dstrain115 reviewed (commented):
Looking good!  I like the very descriptive docstring blocks within the class.

verult commented on cirq-core/cirq/devices/grid_device_metadata.py:
```diff
@@ -169,4 +169,9 @@ def _json_dict_(self):
+        return cls(
+            qubit_pairs,
+            gateset,
+            None if gate_durations is None else dict(gate_durations),
```
Keeping the common behavior up front sgtm, will update

tanujkhattar reviewed (changes requested):
Looking really good. Only concern is around proto validation for asymmetric  and subset permutation targets.

verult commented on cirq-google/cirq_google/devices/grid_device.py:
```diff
@@ -0,0 +1,273 @@
+                gateset=cirq.Gateset(),  # TODO(#5050) implement
+                all_qubits=all_qubits,
+            )
+        except ValueError as ve:  # coverage: ignore
```
I added them because these lines are currently unreachable if everything works correctly, and only act as a last line of defense.

tanujkhattar replied:
Can we add a test which gives a bad input that triggers this condition?

verult replied:
The same conditions that would trigger this exception would also cause `DeviceSpecification` validation to fail, so this is captured by the test `test_grid_device_invalid_device_specification`. If validation is buggy and this exception here is triggered, the test should get an exception but will fail because there's an error pattern mismatch.

Per @MichaelBroughton 's suggestion I added spec validation logic at the top of `from_proto` to make it explicit that errors that would've caused `GridDeviceMetadata` instantiation to fail are due to an invalid `DeviceSpecification`.

Will update the error match patterns to include "Invalid DeviceSpecification" to make sure that validation error messages are matched instead of the GridDeviceMetadata ones, in case error messages accidentally match up in the future between the two.

dstrain115 replied:
Is it even worth having the ValueError catching if unreachable?

verult replied:
I think so - it would protect against future bugs in validation logic. For example, when GridDeviceMetadata gets an additional constraint but validation isn't updated, it's useful to give users the hint that the error they are getting is due to the DeviceSpecification being invalid rather than a mistake they made.

tanujkhattar replied:
I also think we should remove this check since the code path is unreachable. 

We can add an assertion instead if we want to make sure we catch the inconsistencies between validation logic vs new constraints in GridDeviceMetadata but I'm not sure if putting everything under the umbrella of  device specification errors is the right thing to do in this case.

verult commented:
Ready for final approvals!

verult reviewed (commented):
My bad, realized I never sent out my comment replies after pushing the code.

dstrain115 commented on cirq-google/cirq_google/devices/grid_device.py:
```diff
@@ -0,0 +1,282 @@
+
+def _validate_device_specification(proto: v2.device_pb2.DeviceSpecification) -> None:
+    """Validates the DeviceSpecification proto.
+
```
This docstring doesn't really state much beyond the function name itself.  I would either delete the whole thing (since it is a private function) or add more information.  For instance, you could mention what validations are done and any assumptions made.

verult replied:
Ah yeah, since this is private and code comments + error messages already describe what's being validated, I simplified the docstring to just "Raises a ValueError if the DeviceSpecification proto is invalid."

dstrain115 commented on cirq-google/cirq_google/devices/grid_device.py:
```diff
@@ -0,0 +1,282 @@
+    """Validates the DeviceSpecification proto.
+
+    Args:
+        proto: The DeviceSpecification proto to validate.
```
I think putting backquotes around DeviceSpecification will allow it to be linked in the api docs.

dstrain115 reviewed (approved):
LGTM, just a few minor nits.
